simple_salesforce/bulk.py

```diff
@@ -0,0 +1,338 @@
+
+        result = _call_salesforce(url=url, method='POST', session=self.session,
+                                  headers=self.headers,
+                                  data=json.dumps(payload))
+        return result.json(object_pairs_hook=OrderedDict)
+
+    def _get_job(self, job_id):
+        """ Get an existing job to check the status """
+
+        url = "{}{}{}".format(self.bulk_url, 'job/', job_id)
+
+        result = _call_salesforce(url=url, method='GET', session=self.session,
+                                  headers=self.headers)
+        return result.json(object_pairs_hook=OrderedDict)
+
+    def _add_batch(self, job_id, data, operation):
+        """ Add a set of data as a batch to an existing job
+        Separating this out in case of later
+        implementations involving multiple batches
+        """
+
+        url = "{}{}{}{}".format(self.bulk_url, 'job/', job_id, '/batch')
+
+        if operation != 'query':
+            data = json.dumps(data)
+
+        result = _call_salesforce(url=url, method='POST', session=self.session,
+                                  headers=self.headers, data=data)
+        return result.json(object_pairs_hook=OrderedDict)
+
+    def _get_batch(self, job_id, batch_id):
+        """ Get an existing batch to check the status """
+
+        url = "{}{}{}{}{}".format(self.bulk_url, 'job/',
+                                  job_id, '/batch/', batch_id)
+
+        result = _call_salesforce(url=url, method='GET', session=self.session,
+                                  headers=self.headers)
+        return result.json(object_pairs_hook=OrderedDict)
+
+    def _get_batch_results(self, job_id, batch_id, operation):
+        """ retrieve a set of results from a completed job """
+
+        url = "{}{}{}{}{}{}".format(self.bulk_url, 'job/', job_id, '/batch/',
+                                    batch_id, '/result')
+
+        result = _call_salesforce(url=url, method='GET', session=self.session,
+                                  headers=self.headers)
+
+        if operation == 'query':
+            url_query_results = "{}{}{}".format(url, '/', result.json()[0])
+            query_result = _call_salesforce(url=url_query_results, method='GET',
+                                            session=self.session,
+                                            headers=self.headers)
+            return query_result.json()
+
+        return result.json()
+
+    def _bulk_operation(self, object_name, operation, data,
+                        external_id_field=None):
+        """ String together helper functions to create a complete
+        end-to-end bulk API request
+
+        Arguments:
+
+        * object_name -- SF object
+        * operation -- Bulk operation to be performed by job
+        * data -- list of dict to be passed as a batch
+        * external_id_field -- unique identifier field for upsert operations
+        * wait -- seconds to sleep between checking batch status
```

Comment:
Small oversight, documented that parameter but didn't add. Updated. Thanks! 

Most Bulk API jobs that I've run complete in 10-30 seconds, so 5 seemed like a reasonable default to avoid waiting too long. 

I don't have much experience with `Future` objects. Also, running a SFDC Bulk API job in a multithreaded process wouldn't be my first choice. If it were a batch submit with incremental returns of individual records, then I could see the benefit. Anything I'm missing here?